src/csharp/Microsoft.Spark/Sql/Types/DataType.cs

```diff
@@ -39,10 +40,13 @@ public abstract class DataType
             typeof(MapType),
             typeof(StructType) };
 
+        private static string[] s_simpleTypeNormalizedNames = null;
+        private static string[] s_complexTypeNormalizedNames = null;
```

Comment:
Can we replace this with something like
```csharp
        private static Lazy<string[]> s_simpleTypeNormalizedNames =
            new Lazy<string[]>(
                () =>s_simpleTypes.Select(t => NormalizeTypeName(t.Name)).ToArray());

        private static Lazy<string[]> s_complexTypeNormalizedNames =
            new Lazy<string[]>(
                () => s_complexTypes.Select(t => NormalizeTypeName(t.Name)).ToArray());
```
With this you should be able to get rid of 
```csharp
        private static int SimpleTypeIndex(string typeName)
        private static int ComplexTypeIndex(string typeName)
        private static void BuildNormalizedStringMapping()
```

as well as remove the if check in
```csharp
        private static string NormalizeTypeName(Type type)
        {
            if (s_simpleTypeNormalizedNames == null)
```

Reply:
> With this you should be able to get rid of

Even with lazy initialization, how can we get rid of `SimpleTypeIndex` and `ComplexTypeIndex`? We still need to figure out what index the incoming type maps to in the normalized names array, no?

Reply:
Can you call 
`s_simpleTypeNormalizedNames.Value.IndexOf(typeName)` at the callsite instead ?

Reply:
Sure. I was thinking of keeping it a method to add doc comments.